tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1306 @@
+#endif
+
+
+namespace {
+static constexpr char kUriScheme[] = "s3";
+static constexpr char kDotAmazonAwsDotCom[] = ".amazonaws.com";
+static constexpr char kAmzBucketRegionHeader[] = "x-amz-bucket-region";
+}  // namespace
+
+namespace tensorstore {
+namespace {
+namespace jb = tensorstore::internal_json_binding;
+
+auto& s3_bytes_read = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/bytes_read",
+    "Bytes read by the s3 kvstore driver");
+
+auto& s3_bytes_written = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/bytes_written",
+    "Bytes written by the s3 kvstore driver");
+
+auto& s3_retries = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/retries",
+    "Count of all retried S3 requests (read/write/delete)");
+
+auto& s3_read = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/read", "S3 driver kvstore::Read calls");
+
+auto& s3_read_latency_ms =
+    internal_metrics::Histogram<internal_metrics::DefaultBucketer>::New(
+        "/tensorstore/kvstore/s3/read_latency_ms",
+        "S3 driver kvstore::Read latency (ms)");
+
+auto& s3_write = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/write", "S3 driver kvstore::Write calls");
+
+auto& s3_write_latency_ms =
+    internal_metrics::Histogram<internal_metrics::DefaultBucketer>::New(
+        "/tensorstore/kvstore/s3/write_latency_ms",
+        "S3 driver kvstore::Write latency (ms)");
+
+auto& s3_delete_range = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/delete_range",
+    "S3 driver kvstore::DeleteRange calls");
+
+auto& s3_list = internal_metrics::Counter<int64_t>::New(
+    "/tensorstore/kvstore/s3/list", "S3 driver kvstore::List calls");
+
+/// sha256 hash of an empty string
+static constexpr char kEmptySha256[] =
+  "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+
+/// An empty etag which should not collide with an actual payload hash
+static constexpr char kEmptyEtag[] = "\"\"";
+
+/// Adds the generation header to the provided builder.
+bool AddGenerationHeader(S3RequestBuilder * builder,
+                         std::string_view header,
+                         const StorageGeneration& gen) {
+  if (StorageGeneration::IsUnknown(gen)) {
+    // Unconditional.
+    return false;
+  } else {
+    // If no generation is provided, we still need to provide an empty etag
+    auto etag = StorageGeneration::IsNoValue(gen)
+                      ? kEmptyEtag : StorageGeneration::DecodeString(gen);
+
+    builder->AddHeader(absl::StrCat(header, ": ", etag));
+    return true;
+  }
+}
+
+
+std::string payload_sha256(const absl::Cord & cord=absl::Cord()) {
+  SHA256Digester sha256;
+  sha256.Write(cord);
+  auto digest = sha256.Digest();
+  auto digest_sv = std::string_view(
+    reinterpret_cast<const char *>(&digest[0]),
+    digest.size());
+
+  return absl::BytesToHexString(digest_sv);
+}
+
+
+struct S3KeyValueStoreSpecData {
```

Comment:
Add a .rst documentation file including the meaning of the spec data.

Reply:
Add a reminder in the TODO for this.